### PR TITLE
chore(api): Api Exception Alignment

### PR DESCRIPTION
**Pull request by Equartey**
*Description of changes:*
Aligns API exceptions & errors with patterns established in amplify-flutter and [amplify-swift](https://github.com/aws-amplify/amplify-swift/blob/main/Amplify/Categories/API/Error/APIError.swift).

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/api/amplify_api_dart/lib/src/graphql/helpers/graphql_response_decoder.dart [hunk 1]
```diff
@@ -62,7 +62,7 @@ class GraphQLResponseDecoder {
     }
     request.decodePath!.split('.').forEach((element) {
       if (!dataJson!.containsKey(element)) {
-        throw const ApiException(
+        throw const UnknownException(
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api_dart/lib/src/graphql/helpers/graphql_response_decoder.dart:65]: same here

File: packages/api/amplify_api_dart/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 2]
```diff
@@ -535,7 +535,7 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
       _pollResponseTimeout,
       onTimeout: () {
         op.cancel();
-        throw const ApiException('Timeout while polling AppSync.');
+        throw const UnknownException('Timeout while polling AppSync.');
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api_dart/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:538]: maybe `NetworkException` or operation? Doesn't seem unknown.

File: packages/api/amplify_api_dart/lib/src/graphql/web_socket/services/web_socket_service.dart [hunk 3]
```diff
@@ -173,7 +173,7 @@ class AmplifyWebSocketService
           );
         }
 
-        final exception = ApiException(
+        final exception = UnknownException(
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api_dart/lib/src/graphql/web_socket/services/web_socket_service.dart:176]: nit: Maybe operation error? Seems like swift uses for similar circumstance https://github.com/aws-amplify/amplify-swift/blob/main/AmplifyPlugins/API/Sources/AWSAPIPlugin/Operation/AWSGraphQLSubscriptionTaskRunner.swift#L195 though not exactly the same.

File: packages/api/amplify_api_dart/lib/src/graphql/utils.dart [hunk 4]
```diff
@@ -61,7 +61,7 @@ ModelSchema getModelSchemaByModelName(
   // ignore: invalid_use_of_protected_member
   final provider = Amplify.API.defaultPlugin.modelProvider;
   if (provider == null) {
-    throw const ApiException(
+    throw const UnknownException(
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api_dart/lib/src/graphql/utils.dart:64]: Could this be a configuration error?

File: packages/api/amplify_api_dart/lib/src/graphql/utils.dart [hunk 5]
```diff
@@ -76,23 +76,23 @@ ModelSchema getModelSchemaByModelName(
   final schema =
       (provider.modelSchemas + provider.customTypeSchemas).firstWhere(
     (elem) => elem.name == modelName,
-    orElse: () => throw ApiException(
+    orElse: () => throw UnknownException(
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api_dart/lib/src/graphql/utils.dart:79]: Could the rest of these be operation or configuration?

File: packages/amplify_core/lib/src/types/exception/api/operation_error.dart [hunk 6]
```diff
@@ -0,0 +1,19 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import 'package:amplify_core/src/types/exception/api/api_exception.dart';
+
+/// {@template amplify_core.api.api_operation_error}
+/// An in-process operation encountered a processing error
+/// {@endtemplate}
+class ApiOperationError extends ApiException {
```
Comment (ragingsquirrel3) [inline packages/amplify_core/lib/src/types/exception/api/operation_error.dart:9]: nit/question: Should this be named `ApiOperationException`? as it seems most our exceptions end w `*Exception` though I know this is following swift.

File: packages/amplify_core/lib/src/types/exception/error/configuration_error.dart [hunk 7]
```diff
@@ -11,7 +11,7 @@ import 'package:amplify_core/amplify_core.dart';
 /// `Amplify.configure`. If that call succeeds, developers do not need to worry
 /// about future configuration errors.
 /// {@endtemplate}
-class ConfigurationError extends AmplifyError {
+class ConfigurationError extends AmplifyError implements ApiException {
```
Comment (ragingsquirrel3) [inline packages/amplify_core/lib/src/types/exception/error/configuration_error.dart:14]: question: why is this implementing `ApiException`? Isn't it used outside API?

File: packages/amplify_core/lib/src/types/exception/api/invalid_url.dart
Comment (ragingsquirrel3) [inline packages/amplify_core/lib/src/types/exception/api/invalid_url.dart:1]: nit: can the file name have "_exception" at the end?

**Review by ragingsquirrel3 — COMMENTED** [review]
Mostly some nits/questions about naming and which exception to use when/where. Biggest question is regarding keeping `RestException` to keep backwards-compatible there.

File: packages/amplify_core/lib/src/types/exception/error/configuration_error.dart (hunk 7 above)
Comment (dnys1) [reply packages/amplify_core/lib/src/types/exception/error/configuration_error.dart:14]: Shouldn't implement Exception when extending Error

File: packages/amplify_core/lib/src/types/exception/api/operation_error.dart (hunk 6 above)
Comment (dnys1) [reply packages/amplify_core/lib/src/types/exception/api/operation_error.dart:9]: "Error" has a very specific meaning in Dart and is for classes extending `Error` which shouldn't be caught. Recommend `Exception` as well.

File: packages/amplify_core/lib/src/types/api/types/rest/rest_operation.dart [hunk 8]
```diff
@@ -23,7 +23,7 @@ class RestOperation extends AWSHttpOperation<AWSHttpResponse> {
           return response;
         }
         // In case other response types ever added.
-        throw const ApiException('Unable to convert to AWSHttpResponse');
+        throw const ApiOperationError('Unable to convert to AWSHttpResponse');
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/api/types/rest/rest_operation.dart:26]: Better to throw a `StateError` here. Exceptions are used for things which are expected. If we've forgotten to update this code - we messed up and this should fail hard and never be released to prod.

File: packages/amplify_core/lib/src/types/exception/api/http_status_exception.dart [hunk 9]
```diff
@@ -3,19 +3,19 @@
 
 import 'package:amplify_core/amplify_core.dart';
 
-/// {@template rest_exception}
+/// {@template amplify_core.api.http_status_exception}
 /// An HTTP error encountered during a REST API call, i.e. for calls returning
 /// non-2xx status codes.
 /// {@endtemplate}
-class RestException extends ApiException {
+class HttpStatusException extends ApiException {
```
[the remark this answers is not in the document]
Comment (dnys1) [reply packages/amplify_core/lib/src/types/exception/api/http_status_exception.dart:10]: I'm okay with either. If changing name, add a deprecation notice:

```dart
@Deprecated('Use HttpStatusException instead')
typedef RestException = HttpStatusException;
```

File: packages/amplify_core/lib/src/types/exception/error/configuration_error.dart [hunk 10]
```diff
@@ -21,4 +21,11 @@ class ConfigurationError extends AmplifyError {
 
   @override
   String get runtimeTypeName => 'ConfigurationError';
+
+  @override
+  List<Object?> get props => [
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/exception/error/configuration_error.dart:26]: Is this needed?

File: packages/api/amplify_api_dart/lib/src/graphql/helpers/graphql_response_decoder.dart [hunk 11]
```diff
@@ -46,7 +46,7 @@ class GraphQLResponseDecoder {
     // From here, it appears this is a response that must be parsed into a non-string object.
 
     if (request.decodePath == null) {
-      throw const ApiException(
+      throw const UnknownException(
```
[the remark this answers is not in the document]
Comment (dnys1) [reply packages/api/amplify_api_dart/lib/src/graphql/helpers/graphql_response_decoder.dart:49]: Agree operation exception is better

File: packages/api/amplify_api_dart/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 12]
```diff
@@ -274,7 +274,7 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
       // disconnect. Ignore those messages.
       _safeAdd,
       onError: (Object error, StackTrace st) {
-        final exception = ApiException(
+        final exception = UnknownException(
```
Comment (dnys1) [inline packages/api/amplify_api_dart/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:277]: NetworkException

File: packages/api/amplify_api_dart/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart (hunk 2 above)
Comment (dnys1) [reply packages/api/amplify_api_dart/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:538]: I like NetworkException

File: packages/api/amplify_api_dart/lib/src/graphql/web_socket/services/web_socket_service.dart (hunk 3 above)
Comment (dnys1) [reply packages/api/amplify_api_dart/lib/src/graphql/web_socket/services/web_socket_service.dart:176]: Yes, Unknown should be used very sparingly

File: packages/amplify_core/lib/src/types/exception/api/invalid_url.dart [hunk 13]
```diff
@@ -0,0 +1,19 @@
+// Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+import 'package:amplify_core/src/types/exception/api/api_exception.dart';
+
+/// {@template amplify_core.api.api_invalid_url_exception}
+/// Exception thrown when an invalid URL is provided to the request
+/// {@endtemplate}
+class ApiInvalidURLApiException extends ApiException {
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/exception/api/invalid_url.dart:9]: I only see this being used once - is this needed? Or should it be used in more places?

Comment (Equartey) [reply packages/amplify_core/lib/src/types/exception/api/invalid_url.dart:9]: Removed in favor of `ConfigurationError`

File: packages/amplify_core/lib/src/types/exception/api/http_status_exception.dart (hunk 9 above)
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/amplify_core/lib/src/types/exception/api/http_status_exception.dart:13]: Good callout. Going with `HttpStatusException` here as it is more congruent with the underlying error. As such I added the deprecation notice.

File: packages/amplify_core/lib/src/types/exception/api/operation_exception.dart [hunk 14]
```diff
@@ -0,0 +1,19 @@
...
+class ApiOperationException extends ApiException {
+  /// {@macro amplify_core.api.api_operation_error}
+  const ApiOperationException(
+    super.message, {
+    super.recoverySuggestion,
+    super.underlyingException,
+  });
+
+  @override
+  String get runtimeTypeName => 'ApiOperationError';
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/exception/api/operation_exception.dart:18]:
```suggestion
  String get runtimeTypeName => 'ApiOperationException';
```

File: packages/amplify_core/lib/src/types/exception/api/http_status_exception.dart [hunk 15]
```diff
@@ -3,19 +3,22 @@
...
+  HttpStatusException(this.response) : super(response.decodeBody());
 
   /// The HTTP response from the server.
   final AWSHttpResponse response;
 
   @override
   String toString() {
-    return 'RestException{response=$response}';
+    return 'HttpStatusException{response=$response}';
   }
```
Comment (dnys1) [inline packages/amplify_core/lib/src/types/exception/api/http_status_exception.dart:23]: I would think we'd want to print the `response.decodeBody()` instead which is effectively `this.message`, which is already printed without overriding `toString`. I would remove this override and just update `runtimeTypeName` instead.

File: packages/api/amplify_api_dart/lib/src/graphql/providers/oidc_function_api_auth_provider.dart [hunk 16]
```diff
@@ -20,7 +20,7 @@ class OidcFunctionAuthProvider extends TokenAmplifyAuthProvider {
   Future<String> getLatestAuthToken() async {
     final token = await authProvider.getLatestAuthToken();
     if (token == null) {
-      throw const ApiException(
+      throw const UnknownException(
```
Comment (dnys1) [inline packages/api/amplify_api_dart/lib/src/graphql/providers/oidc_function_api_auth_provider.dart:23]: I would make this an `ApiOperationException` or you could introduce an `ApiAuthException` for these

**Review by dnys1 — APPROVED** [review]

File: packages/amplify_core/lib/src/types/exception/api/http_status_exception.dart (hunk 15 above)
Comment (dnys1) [reply packages/amplify_core/lib/src/types/exception/api/http_status_exception.dart:23]: `response.statusCode` should also be printed, I would think, given it's a status exception

**Review by ragingsquirrel3 — APPROVED** [review]

**Pull request merged**